content/blog/there-and-back-again/index.md

```diff
@@ -0,0 +1,505 @@
+Just Pink
+
+Prelude> fromString "Geronimo"
+Nothing
+```
+
+As you can see, in Haskell Land we were able to define the relationship
+between `toString` and `fromString`. If we add a new variant to the `Suit`
+type, it will be picked up and everything will work as intended. This
+technique is described in more detail in the wonderful [Haskell
+Mini-Patterns
+Handbook](https://kowainik.github.io/posts/haskell-mini-patterns#bidirectional-parsing).
+
+## Coming back home
+
+Is there a way to achieve something similar in Elm?
+
+Yes, there is! 🍨
+
+Truth to be told, we can't do _exactly_ what Haskell does, but we can _emulate_ Haskell's behaviour by using a wonderful library called [elm-review](https://package.elm-lang.org/packages/jfmengels/elm-review/latest/).
+
+`elm-review` analyzes your Elm code and reports patterns that violate
+certain rules. We can use it to enforce additional guarantees on our
+projects, such as removing unused functions, unused values, unused modules,
+and unused dependencies. You can read more about it
+[here](https://jfmengels.net/announcing-elm-review) and
+[here](https://jfmengels.net/safe-dead-code-removal).
+
+In our case, we can write a rule that detects functions named like
+`allSuits` and tell it to warn us when we forget to include a type
+constructor. I've shamefully stolen a [gist](https://gist.github.com/jfmengels/e1fd40af3d0e7bde707c0241bf46920f) from Jeroen Engels and pushed it
+as an elm-review package
+[here](https://github.com/Arkham/elm-review-no-missing-type-constructor).
+
+If you'd like to follow along you can pull [this repo](https://github.com/Arkham/elm-there-and-back-again) and then run:
+
+```bash
+$ npx elm-review --template \
+  Arkham/elm-review-no-missing-type-constructor/example
+✔ Fetching template information
+-- ELM-REVIEW ERROR ------------------- src/Main.elm:46:1
+
+NoMissingTypeConstructor: `allSuits` does not contain all
+the type constructors for `Suit`
+
+45| allSuits : List Suit
+46| allSuits =
+    ^^^^^^^^
+47|     [ Green, Yellow, Red, Blue ]
+
+We expect `allSuits` to contain all the type constructors
+for `Suit`.
+
+In this case, you are missing the following constructors:
+, Pink
+
+I found 1 error in 1 file.
+```
+
+🎉 🎉 🎉
+
+Now we can implement a solution which is very similar to the Haskell one!
+
+```elm
+import Dict exposing (Dict)
+
+suitDict : Dict String Suit
+suitDict =
+    Dict.fromList <|
+        List.map
+            (\suit -> ( toString suit, suit ))
+            allSuits
+
+fromString : String -> Maybe Suit
+fromString value =
+    Dict.get value suitDict
+```
+
+Now when we add a new variant of `Suit`, the Elm compiler will nudge us to
+update `toString`, while elm-review will nudge us to add the missing
+variant to `allSuits`, which in turn is going to make sure that `toString`
+and `fromString` will never go out of sync. Here's a link to the [final
+version](https://ellie-app.com/cDrHpYTXsr9a1) of the code for you to
+peruse.
+
+Thanks for reading! 👋
```

Comment:
Probably out of scope of this post, but this got me curious into how the rule works and how I would go about writing my own rules.

Are you planning to write about that? If you are, adding a plug for that next post at the end might be nice.

Reply:
Yes! next post incoming!